common/log.h

```diff
@@ -47,6 +50,332 @@
 //
 // --------------------------------
 
+#ifdef LOG_WITH_TEST
+inline void log_test();
+#endif
+
+// Utility for synchronizing log configuration state
+//  since std::optional was introduced only in c++17
+enum class LogTriState
+{
+    Same,
+    False,
+    True
+};
+
+enum class LogTargetChannel
+{
+    Logfile,
+    Screen
+};
+
+class LogStateWrapper
+{
+    protected:
+        LogStateWrapper(){};
+
+        ~LogStateWrapper()
+        {
+            log_flush_all_targets();
+            for(auto t: _targets){delete t;}
+        }
+
+    private:
+        LogStateWrapper( LogStateWrapper& other ) = delete;
+		void operator=( const LogStateWrapper& )  = delete;
+
+    private:
+        static LogStateWrapper& instance()
+        {
+            static LogStateWrapper inst;
+            return inst;
+        }
+
+        class LogTargetWrapper
+        {
+            private:
+                LogTargetWrapper( LogTargetWrapper& other ) = delete;
+                void operator=( const LogTargetWrapper& )   = delete;
+
+            public:
+                LogTargetWrapper( FILE* handle )
+                :   _type(Type::Stream),
+                    _opened(true),
+                    _handle(handle)
+                {}
+
+                LogTargetWrapper( const std::string && filename )
+                : LogTargetWrapper(filename)
+                {}
+
+                LogTargetWrapper( const std::string & filename )
+                :   _filename(filename)
+                {
+                    if(_filename.empty())
+                    {
+                        _type = Type::Stream;
+                        _handle = stderr;
+                        _opened = true;
+                    }
+                    else
+                    {
+                        _type = Type::File;
+                    }
+                }
+
+                ~LogTargetWrapper()
+                {
+                    if(_type == Type::File && _handle != nullptr) {std::fclose(_handle);}
+                }
+
+                enum class Type{
+                    Invalid,
+                    File,
+                    Stream
+                };
+
+            public:
+                operator FILE*()
+                {
+                    if(!_opened)
+                    {
+                        while( _lock.test_and_set(std::memory_order_acquire) ){ std::this_thread::yield(); }
+                        if(!_opened && _type == Type::File) // check for opened again after acquiring a lock
+                        {
+                            auto result = std::fopen(_filename.c_str(), "w");
+                            if( result )
+                            {
+                                _handle = result;
+                            }
+                            else
+                            {
+                                std::fprintf(stderr, "Failed to open logfile '%s' with error '%s'\n", _filename.c_str(), std::strerror(errno));
```

Comment:
Please keep lines of code under 120 chars wide.

Reply:
Hmmm, `LOG`s in `main.cpp` and `speculative.cpp` aren't formatted to 120 either, should I change it too ?

Reply:
Sure, that wouldn't be a bad idea.